Clang Migration TF Windows

mraunak opened the pull request:
This PR migrates TensorFlow builds on Windows from the MSVC to the clang-cl compiler.

Reason for migration:
Bring TensorFlow Build on Windows and Linux on the same compiler platform to facilitate convenient debugging for the issues related to compiler

Comments:

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/def_file_filter/def_file_filter.py.tpl#L309
Deleted #def_fp.write("\t ??0CoordinatedTask@tensorflow@@qeaa@XZ\n") # for _pywrap_tfe
#Fix the issue of being unable to export the symbol tensorflow::CoordinatedTask::CoordinatedTask(void)
as the symbol was not found and the above command was responsible for exporting the symbol

https://github.com/tensorflow/tensorflow/blob/master/.bazelrc#L410
#clang-cl supports /arch:AVX instead of /arch=AVX

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ir/utils/shape_inference_utils.h#L30
This specific issue is caused by the OpRegistrationData type being defined as [struct OpRegistrationData](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/op_def_builder.h#L67) but declared in failing translation unit as [class OpRegistrationData](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ir/utils/shape_inference_utils.h#L30). Clang apparently makes the difference here (unlike msvc) and it can't find the symbol with class attribute while it is defined only as a struct. Changing [class OpRegistrationData](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ir/utils/shape_inference_utils.h#L30) to struct OpRegistrationData fixes this specific issue.

mraunak committed:
Clang Migration Windows

MichaelHudgins commented on BUILD:
```diff
@@ -4,3 +4,12 @@ exports_files([
     "LICENSE",
 ])
+# Add a platform target to support clang-cl on Windows
+platform(
```
The platform here should not be in the root build file.  Maybe down in tools/toolchains.

mraunak replied:
Done

MichaelHudgins commented on WORKSPACE:
```diff
@@ -86,3 +86,12 @@ tf_workspace1()
 tf_workspace0()
+
+# Register the platform and toolchain to support clang-cl on Windows
+register_execution_platforms(
```
Toolchains should be registered down in https://github.com/tensorflow/tensorflow/blob/master/tensorflow/workspace2.bzl#L94

mraunak replied:
Done

MichaelHudgins reviewed (changes requested):
Can you please note which clang version and any other dependencies that were needed to be installed so internal builds / containers can be updated.

mraunak commented:
> Can you please note which clang version and any other dependencies that were needed to be installed so internal builds / containers can be updated.

clang version: LLVM-15.0.7-win64.exe downloaded from https://github.com/llvm/llvm-project/releases/tag/llvmorg-15.0.7, 
$ set BAZEL_LLVM=<LLVM installation root path> e.g. C:/Program Files/LLVM
$ set Path=<LLVM installation root path>\bin e.g. C:/Program Files/LLVM/bin

mraunak committed:
Resolved Comments

mraunak committed:
added clang-cl build

mraunak committed:
Fix Alignment

mraunak commented:
@MichaelHudgins, all the above-suggested changes have been made. Please review and let me know if any change is needed

angerson commented:
@mraunak What test suites (bazel flags, etc.) have been run against this?

mraunak commented:
> @mraunak What test suites (bazel flags, etc.) have been run against this?

Hi @angerson,

bazel --output_user_root=d:/ --windows_enable_symlinks test ^
--experimental_cc_shared_library --enable_runfiles --nodistinct_host_configuration ^
--dynamic_mode=off --config=xla --config=short_logs --announce_rc ^
--build_tag_filters=-no_pip,-no_windows,-no_oss,-gpu,-tpu --build_tests_only --config=monolithic ^
--config=opt --copt=/std:c++latest --copt=/clang:-Weverything --linkopt=/FORCE:MULTIPLE ^
--extra_execution_platforms=//tensorflow/tools/toolchains/win:x64_windows-clang-cl --compiler=clang-cl ^
--repo_env=TF_PYTHON_VERSION=3.10 -k --test_output=errors ^
--test_tag_filters=-no_windows,-no_oss,-gpu,-tpu --discard_analysis_cache ^
--test_size_filters=small,medium --jobs=16 --test_timeout=300,450,1200,3600 --verbose_failures ^
 -- //tensorflow/... -//tensorflow/java/... -//tensorflow/lite/... -//tensorflow/compiler/... -//tensorflow/python/data/experimental/kernel_tests/... ^
More?   1>run.log 2>&1

vam-google commented on tensorflow/tools/ci_build/windows/cpu/pip/build_tf_windows_clang-cl.sh:
```diff
@@ -0,0 +1,166 @@
+bazel build ${EXTRA_BUILD_FLAGS}  \
+  --experimental_cc_shared_library \
+  --build_tag_filters=-no_pip,-no_windows,-windows_excluded,-no_oss,-oss_excluded,-gpu,-tpu \
+  --copt=/std:c++latest --copt=/clang:-Weverything --linkopt=/FORCE:MULTIPLE \
```
This file seems to be copied from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/windows/cpu/pip/build_tf_windows.sh almost as is, with a few parameters added as this one. Is it possible to avoid this duplication (like having one shell script containing all shared parts, making these extra parameters an attribute to the script?

angerson replied:
We're going to toss scripts from this directory anyway once they're properly integrated into the ci/official directory, which is currently ongoing, so I don't think it's worth the time to add extra logic.

mraunak replied:
Yes, it is possible to avoid the duplication in two ways:

1. Make changes in the .bazelrc file (https://github.com/tensorflow/tensorflow/blob/master/.bazelrc).
Add extra parameters related to clang-cl such as --copt, linkopt, platforms etc 

for this, I need to test if these changes in .bazelrc do not affect the ongoing build on the MSVC platform

2. Pass extra parameters as arguments to the script build_tf_windows.sh
These changes will be internal to Intel :
https://github.com/Intel-tensorflow/tensorflow/blob/mraunak/win_ci_scripts/tensorflow/tools/ci_build/windows/cpu/pip/ci_run_build_test_clang.sh#L144

mraunak replied:
> We're going to toss scripts from this directory anyway once they're properly integrated into the ci/official directory, which is currently ongoing, so I don't think it's worth the time to add extra logic.

Hi @vam-google and @angerson are we okay with https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/windows/cpu/pip/build_tf_windows.sh 
Please let me know if any further change is needed from my end. Thank you for your comments

angerson replied:
If it's possible to add all clang-unique changes to a bazelrc configuration option (like `build:win_clang`), then please go with that.

mraunak replied:
Hi @angerson, added clang-unique changes to the .bazelrc file, please review and let me know if any further change is needed

vam-google reviewed (commented):
Please also add a comment about the current plan to submit it with clang 15, and then move it to 17 skipping 16, as it is important to have linux and windows on the same version of clang.

mraunak commented:
> Please also add a comment about the current plan to submit it with clang 15, and then move it to 17 skipping 16, as it is important to have linux and windows on the same version of clang.

Note: 
Currently, we have migrated the TensorFlow build from MSVC to the Clang-cl compiler with Clang version 15.0.7, it is compatible with the Windows platform unlike Clang 16.0 as we observed a few errors while using Clang 16 on Windows. Next, we will upgrade Clang 15.0.7 to Clang 17 skipping 16 due to compatibility issues

mraunak committed:
remove redundancy

mraunak committed:
Merge branch 'mraunak/Clang_TF_Win' of https://github.com/Intel-tensorflow/tensorflow into mraunak/Clang_TF_Win

mraunak committed:
added clang configs to .bazelrc

mraunak committed:
added host_linkopt

angerson commented on tensorflow/workspace2.bzl:
```diff
@@ -114,6 +114,10 @@ def _tf_toolchains():
     def_file_filter_configure(name = "local_config_def_file_filter")
 
+    # Register the toolchain to support clang-cl on Windows
+    native.register_toolchains("@local_config_cc//:cc-toolchain-x64_windows-clang-cl")
```
This needs to be applied only on Windows and only if clang is enabled. Our other platforms are getting presubmit errors like:

```
ERROR: /[tf/tensorflow/tensorflow/python/compiler/xla/BUILD:26](https://cs.corp.google.com/piper///depot/google3/tf/tensorflow/tensorflow/python/compiler/xla/BUILD?l=26&cl=565719930):20: While resolving toolchains for target 
//tensorflow/python/compiler/xla:jit_test_cpu: invalid registered toolchain '@local_config_cc//:cc-toolchain-x64_windows-clang-cl': no such target '@local_config_cc//:cc-toolchain-x64_windows-clang-cl': target 'cc-toolchain-x64_windows-clang-cl' not declared in package '' defined by 
/root/.cache/bazel/_bazel_root/fbac33eb30dbfb6b11b15a7ff5ac830d/external/local_config_cc/BUILD (Tip: use `query "@local_config_cc//:*"` to see all the targets in that package)
```

mraunak replied:
@angerson, I have made changes to fix the pre-submit error. Please review and validate, and let me know if it still fails or if there is any other failure

mraunak committed:
fix presubmit failure

MichaelHudgins commented on third_party/tf_runtime_clangcl.patch:
Can this file be moved into the tf_runtime subfolder?  TSL and XLA repos copy the folder

MichaelHudgins commented on third_party/tf_runtime_clangcl.patch:
Can this file be moved into the tf_runtime subfolder? TSL and XLA repos copy the folder

mraunak replied:
Done

mraunak committed:
patch_file moved

mraunak committed:
tf_runtime_clangcl.patch in subfolder

mraunak committed:
Create BUILD

//third_party/tf_runtime:tf_runtime_clangcl.patch: BUILD file not found in any of the following directories. Add a BUILD file to a directory to mark it as a package

MichaelHudgins commented on third_party/tf_runtime/BUILD:
If a new empty build file is required it needs to be totally empty (it is causing a somewhat silly lint error) and please add it to the opensource_only.files , this helps keep track of build files not relevant internally.  Probably needs go go above this line https://github.com/tensorflow/tensorflow/blob/master/tensorflow/opensource_only.files#L353

MichaelHudgins replied:
To be clear the one empty line in the file is causing the lint issue

mraunak replied:
Made Changes, please review it and let me know if any further alteration is needed

mraunak committed:
fix the lint issue

mraunak committed:
correction
